### DISCUSSION TITLE
Q&A for projects

### DISCUSSION THREAD

**Discussion by stan6** [SOEN345-WINTER2024/discussion#6] [Announcements]
```
 Is it required to run the project? The requirements state that the project should be compiled. However, even when i compile the project there isn't much i can do with it, even when i try to run its existing tests.

```
As stated in the requirement, you need to compile the project because (1) you may need to write new unit tests for some of the methods and you cannot test a particular method if you cannot compile it, (2) you may need to fix a bug if you find new bugs (you cannot verify the correctness of the fix if you cannot even compile it.

For compilation of open-source projects, Android projects usually use maven or gradle for building. These projects usually can be easily compiled by running mavenw and gradlew which are wrapper for running the build/run tests command. You should also read carefully the README of a project about instructions in building and running the project as this information is usually clearly written.

**Comment by stan6** [comment]
Q: How to calculate line of code for my project?
A: You can use the API of the codetab:
https://api.codetabs.com/v1/loc?github=jolav/betazone
For example, for the repo: https://github.com/jolav/betazone, the link https://api.codetabs.com/v1/loc?github=jolav/betazone will lead to the JSON file of calculating.

**Comment by stan6** [comment]
Q: ` I feel a bit lost and I just wanted to know in between integration testing & unit testing what would be the difference in testing.`

A: Below are the kinds of testing that you can do for an app:
- Write JUnit tests for some selected methods in the app
- Integration testing by calling several methods in the app. This is a bit difficult as you will need to find a few related methods that can be called together. 
- Manually testing the app by playing with the selected app in the device. We also talk about Android testing via record-and-replay using Espresso in Lab 3

Please note that integration testing is not required and when you find a few methods for unit testing, you will naturally need to call a few methods (e.g., constructors) to construct the object and to check the correctness of the output. In general, you will mainly do manual testing by yourself. After we taught about several ways of modeling, you will design unit tests based on the learned criteria.

**Comment by stan6** [comment]
Q: Our team is wondering how we are supposed to submit the proposal. Is it 3 documents, 1 for each app, or is it one document?
A: There should be one document where the answers for all apps are included in the document README.md.

**Comment by gevorgalaverdyan** [comment]
Hello, is it ok that instead of Google Play, my app is in F-Droid ?

**Reply by stan6** [reply]
Yes, as long as it is hosted on GitHub. You can add the note about it being in F-Droid.

**Comment by eimcharles** [comment]
Hello, i found an android project https://github.com/SimpleMobileTools/Simple-Gallery that meets all the criterion. However i cannot seem to find any test classes associated within the repository. Is this okay?

**Reply by stan6** [reply]
Yes, you just need to answer the question about test in your README.

**Reply by eimcharles** [reply]
Awesome ! Thank you

**Comment by abdelh17** [comment]
I see we need to use Monkey Runner, are there resources/tutorials you recommend we use?

**Reply by stan6** [reply]
We mentioned about Espresso (https://developer.android.com/training/testing/espresso) or manual testing in the instruction. All of these are for recording the UI test and replaying.

Regards,
Shin Hwei

**Comment by stan6** [comment]
> For the Event-flow Graphs (EFG), do you want us to draw the EFG of all the packages that we have? 

> Like here, I have to check all the classes in each package and draw the EFG? 

The EFG is drawn by referring to the GUI not the classes so you should draw the EFG by looking at from the main screen of your app, what events (e.g., button to click, text field to fill in) can be performed and then follow by the next screen after certain event. You can refer to https://www.cs.umd.edu/~atif/GUITAR-Web/guitar_process_efg.htm for example EFG.

**Comment by stan6** [comment]
Q:for the final presentation, I don't understand what should I do for this: 
"
◼ Timeline of the project (Use the following format for all the GitHub Classroom submissions that are related to testing your Android apps, including MPs, Project Proposal, Project Progress Report, and Lab): (2 points) 
◆ Stage: Project Proposal 
◆ Date Submitted: 
◼ Goals Achieved (e.g., tested how many methods, found how many bugs):"


A: Hi, I don't understand what is confusing here. It gave clear examples for each part you need to write. There should be a timeline slide with several stages (one example stage is the project proposal) and you should write when is the project proposal submitted for "date submitted" and goal achieved during that stage.

**Comment by stan6** [comment]
Q: 
I didn’t really understand what you meant by “experimental results demonstrating the effective of testing techniques used” in the presentation description. Is it just explaining how we found the bugs using the appropriate testing technique?

A:The slides are designed based on template of a research presentation where you have shown in earlier slide that at each stage of timeline, you have used some testing techniques. Naturally, you need to present results about these techniques. We include detailed examples about the results slide (e.g., efficiency like time taken by each testing technique and the number of each bug found by each technique). There are also examples of the technique (e.g., JUnit, collaborative bug finding using GitHub issues).
